### ISSUE TITLE
Configurable struct tags

### ISSUE THREAD

**Issue by schmichael** [zclconf/go-cty#6]
For [Nomad](https://github.com/hashicorp/nomad) we'd love to be able to configure the struct field tag cty uses like the [BasicHandle.TypeInfos](https://godoc.org/github.com/ugorji/go/codec#BasicHandle) field in `codec`.

With `codec` we can configure our codec with:
```go
	h := &codec.MsgpackHandle{}
	h.TypeInfos = codec.NewTypeInfos([]string{"cty", "codec"})
```
...to enable struct field tag compatibility.

Ideally we could set the struct field tag on *both* to `hcl` to avoid confusion for users implementing plugins.

### Example

With zero tag overriding our struct fields would need both tags:
```go
	RunFor time.Duration `cty:"run_for" codec:"run_for"`
```

With codec's we can at least reduce that number to 1 tag:
```go
	RunFor time.Duration `cty:"run_for"`
```

However "cty" means nothing to Nomad users, so replacing it with "hcl" seems optimal:
```go
	RunFor time.Duration `hcl:"run_for"`
```

**Comment by apparentlymart** [comment]
Hmm... this feels a little tricky to me because the format of the struct tags is not identical between the two... the `hcl` struct tags have some extra modifiers to choose between block/attribute/etc while the `cty` ones don't.

It honestly feels weird to me to expose _either_ of these directly to a plugin developer; that seems like some very tight coupling between the plugin API and Nomad's current choice of configuration syntax.

For context, over in Terraform we're using `cty/msgpack` within the plugin SDK primarily for convenience (since it's a msgpack interface we already know well) but the plugin developer using our SDK never knows about it and, if a developer were not using our SDK (i.e. in future writing a plugin in a different language entirely) it'd be fine for them to just use any old msgpack parser/formatter as long as it fully supports the latest spec.

(We also added an extra higher-level schema layer over the low-level ones in HCL/cty, and while we did that primarily because Terraform needs some additional metadata that the lower-level systems can't represent, it also serves to insulate plugins from these configuration language details in the _other_ direction.)
